helix-core/src/object.rs

```diff
@@ -1,76 +1,74 @@
-use crate::{Range, RopeSlice, Selection, Syntax};
-use tree_sitter::Node;
+use crate::{syntax::TreeCursor, Range, RopeSlice, Selection, Syntax};
 
 pub fn expand_selection(syntax: &Syntax, text: RopeSlice, selection: Selection) -> Selection {
-    select_node_impl(syntax, text, selection, |mut node, from, to| {
-        while node.start_byte() == from && node.end_byte() == to {
-            node = node.parent()?;
+    let cursor = &mut syntax.walk();
+
+    selection.transform(|range| {
+        let from = text.char_to_byte(range.from());
+        let to = text.char_to_byte(range.to());
+
+        let mut byte_range = from..to;
+        cursor.reset_to_byte_range(from, to);
+
+        loop {
+            if cursor.node().byte_range() != byte_range {
+                break;
+            }
+            byte_range = cursor.node().byte_range();
+            if !cursor.goto_parent() {
+                break;
+            }
         }
```

Comment:
```suggestion
        while cursor.node().byte_range() == byte_range {
        	if !cursor.goto_parent() {
        		break
        	}
        }
```

this can be simplified.

`byte_range` doesn't need to be mutable since its never read after the loop and the loop arborts if the byte range changes.

Then you can use a while loop (you could fold the goto_parent into the while condition but I find loops with empty bodies weird)